essentials-framework/Essentials Core/PepperDashEssentialsBase/Utilities/ActionSequence.cs

```diff
@@ -0,0 +1,153 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using Crestron.SimplSharp;
+using Crestron.SimplSharpPro.CrestronThread;
+
+using PepperDash.Core;
+using PepperDash.Essentials.Core;
+using PepperDash.Essentials.Core.Config;
+
+using Newtonsoft.Json;
+
+namespace PepperDash.Essentials.Core.Utilities
+{
+    /// <summary>
+    /// A device that executes a sequence of actions with optional delays between actions
+    /// </summary>
+    [Description("A device that exectues a sequence of actions with optional delays between actions")]
+    public class ActionSequence : EssentialsDevice
+    {
+        private ActionSequencePropertiesConfig _propertiesConfig;
+
+        private CrestronQueue<SequencedDeviceActionWrapper> _actionQueue;
+
+        private Thread _worker;
+
+        private bool _allowActionsToExecute;
+
+        public ActionSequence(string key, DeviceConfig config)
+            : base(key, config.Name)
+        {
+            var props = config.Properties.ToObject<ActionSequencePropertiesConfig>();
+            _propertiesConfig = props;
+
+            if (_propertiesConfig != null)
+            {
+                if (_propertiesConfig.ActionSequence.Count > 0)
+                {
+                    _actionQueue = new CrestronQueue<SequencedDeviceActionWrapper>(_propertiesConfig.ActionSequence.Count);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Starts executing the sequenced actions
+        /// </summary>
+        public void StartSequence()
+        {
+            Debug.Console(1, this, "Starting Action Sequence");
+            _allowActionsToExecute = true;
+            AddActionsToQueue();
+            _worker = new Thread(ProcessActions, null, Thread.eThreadStartOptions.Running);
```

Comment:
We might want to check if the thread is already running and abort the start sequence if it is.

Reply:
Either that, or put the whole while loop in the `ProcessActions` method into a CrticalSection so that only one thread can be pulling things off the queue at a time, although I'm betting we'll get an exception if you try and create a new thread while the old one is still running.

Reply:
good call.  I'll add check